hack/generator/pkg/codegen/pipeline_export_generated_code.go

```diff
@@ -0,0 +1,197 @@
+/*
+ * Copyright (c) Microsoft Corporation.
+ * Licensed under the MIT license.
+ */
+
+package codegen
+
+import (
+	"context"
+	"fmt"
+	"github.com/Azure/k8s-infra/hack/generator/pkg/astmodel"
+	"github.com/pkg/errors"
+	"k8s.io/klog/v2"
+	"os"
+	"path/filepath"
+	"sort"
+)
+
+// exportPackages creates a PipelineStage to export our generated code as a set of packages
+func exportPackages(outputPath string) PipelineStage {
+	description := fmt.Sprintf("Export packages to %q", outputPath)
+	return PipelineStage{
+		description,
+		func(ctx context.Context, types map[astmodel.TypeName]astmodel.TypeDefiner) (map[astmodel.TypeName]astmodel.TypeDefiner, error) {
+			packages, err := CreatePackagesForDefinitions(types)
+			if err != nil {
+				return nil, errors.Wrapf(err, "failed to assign generated definitions to packages")
+			}
+
+			packages, err = MarkLatestResourceVersionsForStorage(packages)
+			if err != nil {
+				return nil, errors.Wrapf(err, "unable to mark latest resource versions for as storage versions")
+			}
+
+			err = writeFiles(ctx, packages, outputPath)
+			if err != nil {
+				return nil, errors.Wrapf(err, "unable to write files into %q", outputPath)
+			}
+
+			return types, nil
+		},
+	}
+}
+
+// CreatePackagesForDefinitions groups type definitions into packages
+func CreatePackagesForDefinitions(definitions map[astmodel.TypeName]astmodel.TypeDefiner) ([]*astmodel.PackageDefinition, error) {
+
+	genVersion := combinedVersion()
+	packages := make(map[astmodel.PackageReference]*astmodel.PackageDefinition)
+	for _, def := range definitions {
+		defName := def.Name()
+		groupName, pkgName, err := defName.PackageReference.GroupAndPackage()
+		if err != nil {
+			return nil, err
+		}
+
+		pkgRef := defName.PackageReference
+		if pkg, ok := packages[pkgRef]; ok {
+			pkg.AddDefinition(def)
+		} else {
+			pkg = astmodel.NewPackageDefinition(groupName, pkgName, genVersion)
+			pkg.AddDefinition(def)
+			packages[pkgRef] = pkg
+		}
+	}
+
+	var pkgs []*astmodel.PackageDefinition
+	for _, pkg := range packages {
+		pkgs = append(pkgs, pkg)
+	}
+
```

Comment:
I think we had a "bug" in the original version of this code that this set of packages isn't sorted.

Since maps don't have a defined iteration order, this results in us generating different packages first.

I think we should add:
```
sort.Slice(pkgs, func(i int, j int) bool {
		return pkgs[i].GroupName < pkgs[j].GroupName || (pkgs[i].GroupName == pkgs[j].GroupName && pkgs[i].PackageName < pkgs[j].PackageName)
	})
```

Reply:
Partitioning things into packages occurs earlier, and once that happens they're independent - so do we need to worry about which order they're written?

Reply:
The order which the packages are in in this collection impacts the order they're written to disk right? I think that would show up if something were say, causing a panic during package writing you'd end up with a different set of stuff written to disk every time.

I _think_ if this is ordered, you'd get determinism even there.

Reply:
I'm not 100% convinced this is needed - but it certainly won't hurt. Done.